src/app/courses/courses-list/courses-list.component.ts

```diff
@@ -1,62 +1,102 @@
-import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
+import { Component, OnInit, Input, OnChanges, OnDestroy } from '@angular/core';
+import { debounceTime } from 'rxjs';
 
-import { Course } from 'src/app/shared/interfaces/course.interface.';
-import { FilterPipe } from 'src/app/shared/pipes/filter.pipe';
+import { HttpCourse } from 'src/app/shared/interfaces/http-course.interface';
 import { CourseService } from 'src/app/shared/services/course.service';
 
 @Component({
   selector: 'app-courses-list',
   templateUrl: './courses-list.component.html',
   styleUrls: ['./courses-list.component.css']
 })
-export class CoursesListComponent implements OnInit {
-
-  constructor(private coursesService: CourseService) {}
+export class CoursesListComponent implements OnInit, OnChanges {
 
   @Input() searchValue: string = '';
-
-  courses: Course[] = [];
-  courseForDelete?: Course;
-  titleForDelete = "";
+  courseForDelete?: HttpCourse;
+  titleForDelete: string = '';
   showModal = false;
+  listOfCourses: HttpCourse[] = [];
+  inputValue = '';
+  debounceTimeValue = 500;
 
-
-  ngOnInit() {
-      this.courses = this.coursesService.getListCourses();
+  constructor(private courseService: CourseService ) {
+    
   }
-
-  trackByFn(index: number, item: Course): number {
-    return item.id; // Assuming each item has a unique "id" property
+  
+  ngOnInit() {
+    this.courseService.loadingSubject.next(true);
+    this.courseService.onFetchCourses()
+    .subscribe(
+      courses => {
+        this.listOfCourses = [...courses];
+      }
+    )
+    this.courseService.loadingSubject.next(false);
   }
-
-  filterItems(courses: Course[], searchtext: string): Course[] {
-    let filterPipe = new FilterPipe();
-    return filterPipe.transform(courses, searchtext);
+    
+  ngOnChanges() {
+    this.courseService.inputSubject
+    .pipe(debounceTime(this.debounceTimeValue))
+    .subscribe(value => {
+      this.inputValue = value;
+      if (this.inputValue.length >= 3) {
+        this.courseService.loadingSubject.next(true);
+        this.courseService.onFetchCoursesWithFilter(this.inputValue).subscribe(
+          courses => {
+            this.listOfCourses = [...courses];
+          }
+        )
+        this.courseService.loadingSubject.next(false);
+      } else {
+        this.courseService.loadingSubject.next(true);
+        this.courseService.onFetchCourses().subscribe(
+          courses => {
+            this.listOfCourses = [...courses];
+          }
+        )
+        this.courseService.loadingSubject.next(false);
+      }
+    })
   }
 
   onClickLoadMore(): void {
-    console.log('clicked');
+    this.courseService.onLoadAdditionalCourses();
+    this.courseService.onFetchCourses().subscribe(
+      course => {
+        this.listOfCourses = [...course];
```

Comment:
You can try to create the subject with the courses in the service and subscribe to it 
Only once in the _ngOnInit_ hook
There is no point in so many subscriptions - it breaks the logic of the Angular approach itself

Reply:
Done, created the subject in the service, and redone the logic according to the subscription to the subject with courses